### PR TITLE
Add support for hashing in a worker thread

### LINKED ISSUE #16: Add support for doing the hashing in a worker thread
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/hasha/issues/16)

Could maybe be useful for hashing large files, as the crypto Node.js module is synchronous, so it could block the main thread for some time.

https://nodejs.org/api/worker_threads.html

Thoughts?

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $80.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)
### Submitted pull Requests
- [#21 Move file hashing to thread, solves #16](https://issuehunt.io/r/sindresorhus/hasha/pull/21)
- [#22 Add support for hashing in a worker thread](https://issuehunt.io/r/sindresorhus/hasha/pull/22)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/hasha/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by pro-src**
Fixes #16, Closes #21 

Hi @sindresorhus,

I've seen your code everywhere, in everything, and it's always top notch! And I've seen the billion PRs that you have to review so I really hope that this is as ready as I think it is... I really appreciate your time and look forward to your review.

~Thanks

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#16: Add support for doing the hashing in a worker thread](https://issuehunt.io/repos/36116052/issues/16)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -107,6 +143,50 @@ Returns a `Promise` for the calculated file hash.
...
+
+Returns a `Promise` for the calculated file hash.
+
+### thread.[terminate](https://nodejs.org/api/worker_threads.html#worker_threads_worker_terminate)\()
+
+Stop all JavaScript execution in the worker thread as soon as possible
+
+### thread.[on](https://nodejs.org/api/events.html#events_emitter_on_eventname_listener)\(event, listener)
+
+Adds a new event listener to the worker thread. See [Events](https://nodejs.org/api/worker_threads.html#worker_threads_worker_terminate).
```
Comment (pro-src) [inline readme.md]: Wrong URL, I meant for it to be: https://nodejs.org/api/worker_threads.html#worker_threads_event_error

**Comment by pro-src** [comment]
@sindresorhus 

Changes from previous [commit](https://github.com/sindresorhus/hasha/commit/94ee301a00a579ce1f2ac1eff2e8b40cd32080f5):
* Updated worker code and moved it to *worker.js*, created a class, etc.
* Updated worker typescript definitions and moved them to *worker.d.ts*
* Created *worker.test.d.ts*
* Removed circular references and the `fromString` method from `hasha`
* Renamed `thread.fromString` -> `worker.fromInput`
* Fixed *test.js* node version check, added semver to dev-dependencies
* Fixed travis-ci build, was failing after I fixed the version check
* Updated README to reflect the above changes
* *hasha.js* no longer imports *worker.js* by default.
* Simplified everything, minimal solution.

So, basically I changed everything. :)

**Comment by sindresorhus** [comment]
While I appreciate the PR, I'm not sure why you did it when you knew #21 was already open? I also prefer the approach in #21, as the fact that it's using a worker underneath is just an implementation detail and the user can just use the normal API and benefit from it.

**Comment by pro-src** [comment]
@sindresorhus Considering that it had been sitting for months with changes requested, I thought it had been abandoned.

Not cool.

**Comment by pro-src** [comment]
Just so we're clear, if there's a PR that means it's off limits and I shouldn't bother and/or I should open a PR prematurely to claim an issue for months?

**Comment by sindresorhus** [comment]
@pro-src There's alway less activity during the summer. Personally, I was taking a break from reviewing PRs and @stroncium had missed the notification about the review. I'm sorry that you wasted your time on this, but you could also just have asked before starting the work instead of assuming. Even if @stroncium had abandoned the PR, the best move would have been to continue where he left off as that PR was the intended direction and had already gotten a lot of discussion.

**Comment by pro-src** [comment]
@sindresorhus You're right. It annoyed me that the reason why I opened the PR wasn't clear. I thought it was obvious and that the code here is great. I will follow that sound advice and maybe I'll be able to solve some other issues in the near future.
